js/ui/status/paygo.js

```diff
@@ -0,0 +1,129 @@
+        this._paygNotifier = new Payg.PaygNotifier();
+        this._item.menu.addAction(_("Apply PayGo credit code"), () => {
+                this._paygNotifier.notify(-1);
+        });
+        this.menu.addMenuItem(this._item);
+
+        // show this status applet if PayGo is enabled and fill in
+        // "determining time..." label and icon
+        this._sync();
+
+        // fill in current values as soon as _paygManager is initialized
+        this._updateTimeRemainingAndSyncWhenReady();
+
+        // update immediately when the user extends their time (so they don't
+        // have to wait for the up to REFRESH_TIME_SECS seconds which would
+        // likely be long enough that they might worry something went wrong)
+        this._expiryTimeChangedId = this._paygManager.connect('expiry-time-changed', () => {
+            this._updateTimeRemainingAndSyncWhenReady();
+        });
+
+        // refresh the displayed icon and "time remaining" label periodically
+        this._timeoutRefreshId = Mainloop.timeout_add_seconds (REFRESH_TIME_SECS, () => {
+            this._timeoutRefresh();
+        });
+        GLib.Source.set_name_by_id(this._timeoutRefreshId, '[gnome-shell] this._timeoutRefresh');
+    },
+
+    _timeoutRefresh() {
+        this._updateTimeRemainingAndSyncWhenReady();
+        this._timeoutRefreshId = 0;
+
+        return GLib.SOURCE_CONTINUE;
+    },
+
+    _onDestroy() {
+        this.parent();
+
+        if (this._expiryTimeChangedId != 0)
+            this._paygManager.disconnect(_expiryTimeChangedId);
+
+        if (this._timeoutRefreshId != 0)
+            Mainloop.source_remove(this._timeoutRefreshId);
+    },
+
+    _getMenuGicon() {
+        const URGENT_EXPIRATION_S = 15 * 60;
+        let timeLeftSeconds = this._paygManager.timeRemainingSecs();
+
+        let iconUri = 'resource:///org/gnome/shell/theme/paygo-normal-symbolic.svg';
+        // if time left <= 0, we haven't yet determined it, so fall back to
+        // "normal" icon
+        if (timeLeftSeconds >= 0 && timeLeftSeconds <= URGENT_EXPIRATION_S) {
+            iconUri = 'resource:///org/gnome/shell/theme/paygo-near-expiration-symbolic.svg';
+        }
+
+        return new Gio.FileIcon({ file: Gio.File.new_for_uri(iconUri) });
+    },
+
+    _getTimeRemainingString() {
+        // the time will be invalid if the manager hasn't been
+        // intitialized yet so return with a default message in that case
+        if (!this._paygManager.initialized)
+            return _("Getting time…");;
+
+        let seconds = this._paygManager.timeRemainingSecs();
+        if (seconds < 60)
+            return _("Less than 1 minute");
+
+        return Payg.timeToString(seconds);
+    },
+
+    _sync() {
+        let sensitive = !Main.sessionMode.isLocked && !Main.sessionMode.isGreeter && this._paygManager.enabled;
+        this.menu.setSensitive(sensitive);
+        this._item.actor.visible = this._indicator.visible = this._paygManager.enabled;
```

Comment:
We automatically hit `_sync()` when the `expiry-changed` signal fires. Since `enabled == false` at that point, we'll disable the applet.

So, we aren't strictly keyed to a change in the value of `enabled` changing, I guess, but this should handle every case where it matters.

Or is there something else you think needs to be addressed?